From bf3e7aec47504c75c176542850e917e29465f41d Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <uwu@icenowy.me>
Date: Fri, 8 Mar 2024 17:24:55 +0800
Subject: [PATCH 6/9] Keep unsigned 32-bit parameter sign-extended in register.

On riscv64, 32-bit value in 64-bit register should be sign-extended, even
when it's type is unsigned.

Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
---
 xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp b/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp
index ddd6692156..ee2f7ff767 100644
--- a/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp
+++ b/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp
@@ -48,7 +48,9 @@ extern "C" void invoke_copy_to_stack(uint64_t* gpregs, double* fpregs,
           value = s->val.u16;
           break;
         case nsXPTType::T_U32:
-          value = s->val.u32;
+          // 32-bit values need to be sign-extended in 64-bit registers,
+          // so use the signed value here.
+          value = s->val.i32;
           break;
         case nsXPTType::T_U64:
           value = s->val.u64;
-- 
2.43.4

